frontend/packages/topology/src/data-transforms/DataModelProvider.tsx

```diff
@@ -19,58 +17,104 @@ interface DataModelProviderProps {
+  return null;
+};
+
+export const DynamicWatchResourcesLoader: React.FC<{
+  children: (props: WatchK8sResources<any>) => React.ReactNode;
+}> = ({ children }) => {
+  const [watchResources, setWatchResources] = React.useState<WatchK8sResources<any>>({});
+  const [modelFactories, loaded] = useResolvedExtensions<DynamicTopologyDataModelFactory>(
+    isDynamicTopologyDataModelFactory,
+  );
+  const modelResources = React.useMemo(
+    () =>
+      modelFactories
+        ?.map((m) => ({ id: m.properties.id, resourceHook: m.properties.resources }))
+        .filter(({ resourceHook }) => !!resourceHook),
+    [modelFactories],
+  );
+
+  const onResolve = React.useCallback(
+    (resources: WatchK8sResources<any>) =>
+      setWatchResources((prevResources) => ({ ...prevResources, ...resources })),
+    [],
+  );
+
+  return (
+    <>
+      <>
+        {loaded &&
+          modelResources.map(({ id, resourceHook }) => (
+            <WatchResourcesHookResolver
+              key={id}
+              useResources={resourceHook}
+              onResolve={onResolve}
+            />
+          ))}
+      </>
+      {children(watchResources)}
+    </>
+  );
+};
 
-              return { [k]: { namespace, kind, ...v?.opts } };
-            }),
-          ),
-      },
-    };
+const omitResources = (
+  factories: ResolvedExtension<DynamicTopologyDataModelFactory>[],
+): ResolvedExtension<Omit<DynamicTopologyDataModelFactory, 'resources'>>[] => {
+  const otherProperties = factories.map(({ properties, ...ext }) => {
+    const { resources, ...other } = properties;
+    return { properties: other, ...ext };
   });
+  return otherProperties;
+};
 
 const DataModelProvider: React.FC<DataModelProviderProps> = ({ namespace, children }) => {
   const [model, setModel] = React.useState<ExtensibleModel>(new ExtensibleModel(namespace));
+  const [dataModelFactories, setDataModelFactory] = React.useState<
+    ResolvedExtension<TopologyDataModelFactory | DynamicTopologyDataModelFactory>[]
+  >();
 
   React.useEffect(() => {
     setModel(new ExtensibleModel(namespace));
   }, [namespace]);
 
-  const modelFactories = useExtensions<TopologyDataModelFactory>(isTopologyDataModelFactory);
-  const dynamicModelFactories = useExtensions<DynamicTopologyDataModelFactory>(
-    isDynamicTopologyDataModelFactory,
+  const [modelFactories, modelFactoriesLoaded] = useResolvedExtensions<TopologyDataModelFactory>(
+    isTopologyDataModelFactory,
   );
+  const [dynamicModelFactories, dynamicFactoriesLoaded] = useResolvedExtensions<
+    DynamicTopologyDataModelFactory
+  >(isDynamicTopologyDataModelFactory);
 
-  const namespacedDynamicFactories = React.useMemo(
-    () => getNamespacedDynamicModelFactories(dynamicModelFactories),
-    [dynamicModelFactories],
-  );
+  React.useEffect(() => {
+    if (modelFactoriesLoaded && dynamicFactoriesLoaded) {
+      setDataModelFactory([...modelFactories, ...omitResources(dynamicModelFactories)]);
```

Comment:
Just curious, why do you need to omit resources from the extension? Does this component `DataModelExtension` expect that?

Reply:
@rohitkrai03 idea was not to contribute resources prop twice, so it was omitted from dynamicModelFactories since it's a hook in Dynamic extensions and resolved hook values are contributed through DynamicWatchResourceLoader

Reply:
@rottencandy you can use useMemo here instead of useEffect and useState

Reply:
Updated.